### PR TITLE
Enable 2.1 official builds in AzDO

### PR DESCRIPTION
**Pull request by wtgodbe**
Migrates our TeamCity build infrastructure for release/2.1 official builds to Azure DevOps. 

Notes:

- Pipebuild PR to point the new definition: [here](https://devdiv.visualstudio.com/DevDiv/_git/DotNet-BuildPipeline/pullrequest/294879)
- Pipebuild PR I'll need to merge once I'm done testing end-to-end builds: [here](https://devdiv.visualstudio.com/DevDiv/_git/DotNet-BuildPipeline/pullrequest/294881)
- Aspnet/Buildtools PR to allow customization of inputs to the SignFiles target: [here](https://github.com/aspnet/BuildTools/pull/989), [here](https://github.com/aspnet/BuildTools/pull/990)
- I removed usage of the `aspnet-internal-tools` repo entirely, opting instead to use microbuild to do signing at the end of the Windows-SharedFx job (instead of a separate code signing job that checked out and built aspnet-internal-tools)
- Green test build ran in isolation: [here](https://dev.azure.com/dnceng/internal/_build/results?buildId=930251&view=results)
- Test build with pipebuild inputs, still running: [here](https://dev.azure.com/dnceng/internal/_build/results?buildId=930884&view=results)

Also called out some relevant things in comments on the file changes, and I still have some general cleanup to do, so this is still a WIP.

CC @Pilchie

### CODE REVIEW

File: .azure/pipelines/ci.yml [hunk 1]
```diff
@@ -7,22 +7,661 @@ pr:
...
+      displayName: Download SharedFx installer artifacts
+      inputs:
+        artifactName: artifacts-SharedFx-Installers
+        downloadPath: $(Build.StagingDirectory)/downloaded_artifacts/
+    - task: DownloadBuildArtifacts@0
+      displayName: Download Windows installer artifacts
+      inputs:
+        artifactName: artifacts-Windows-Installers
+        downloadPath: $(Build.StagingDirectory)/downloaded_artifacts/
+    - task: CopyFiles@2
```
Comment (wtgodbe) [inline .azure/pipelines/ci.yml:564]: This giant block of CopyFile tasks is really ugly, but I couldn't get it to work with fewer - @riarenas @mmitche is there a cleaner way to copy all of the inputs without doing them one-by-one?

File: src/Installers/Windows/AspNetCoreModule-Setup/build/settings/common.props [hunk 2]
```diff
@@ -16,8 +16,8 @@
     <IisOobWinSdkVersion Condition="'$(IisOobWinSdkVersion)' == ''">10.0.17134.0</IisOobWinSdkVersion>
     <WindowsTargetPlatformVersion Condition="'$(WindowsTargetPlatformVersion)' == ''">$(IisOobWinSdkVersion)</WindowsTargetPlatformVersion>
     <CharacterSet>Unicode</CharacterSet>
-    <OutDir>bin\$(Configuration)\$(PlatformShortname)\</OutDir>
-    <IntDir>obj\$(Configuration)\$(PlatformShortname)\</IntDir>
+    <OutDir Condition=" '$(OutDir)' == '' ">bin\$(Configuration)\$(PlatformShortname)\</OutDir>
```
Comment (wtgodbe) [inline src/Installers/Windows/AspNetCoreModule-Setup/build/settings/common.props:19]: @JunTaoLuo @dougbu do you remember why we need this setting here at all? This file is imported very late, and up until this moment `outdir` points to `bin\Release\Installers`. Signing fails without this change because the files to sign must be in `outdir`, and this setting changes where `outdir` points to. My change didn't affect where installers got built to, but I'm wondering if I'm missing something.

File: src/Installers/Windows/Wix.props [hunk 3]
```diff
@@ -1,7 +1,13 @@
 <Project>
 
   <PropertyGroup>
-    <Version>$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion).$(AspNetCorePatchVersion).$(BuildNumber)</Version>
+    <!-- Ugly hack - Wix doesn't accept build numbers larger than 65535, and AzDo's BuildIds are larger than that.
```
Comment (wtgodbe) [inline src/Installers/Windows/Wix.props:4]: This hack kinda sucks, but I couldn't think of anything better - @joeloff @dougbu @JunTaoLuo any ideas?

Comment (joeloff) [reply src/Installers/Windows/Wix.props:4]: I think in TC the build number was literally just a counter. If ```BuildNumber``` is used for MSIs, then it doesn't really matter that much because Windows Installer only cares about the first 3 parts. For standalone bundles (EXEs), the 4 part version is considered when evaluating upgrades so having an increasing number is desired.

Are the AzDo numbers incremental as well? Unless it hits something large, e.g. 262140 and the modulus wraps, I guess this would be fine.

Comment (wtgodbe) [reply src/Installers/Windows/Wix.props:4]: Yeah, AzDO numbers are also incremental. We feed the version into `GuidInputs` as of https://github.com/dotnet/aspnetcore/pull/27079. Since the max for any part of the 4-part version, we'd run into possible issues as soon as any build number got past 65535 - it just might happen sooner in AzDO since builds are more frequent there than in TeamCity. But theoretically, the odds of a collision should be the same as the previous odds of an overflow

Comment (wtgodbe) [reply src/Installers/Windows/Wix.props:4]: And we'll always be incrementing the 3rd part of the 4-part version anyways (each patch)

**Comment by Pilchie** [comment]
Not going to claim I'm capable of reviewing this, but very happy to see it.

File: .azure/pipelines/jobs/default-build.yml [hunk 4]
```diff
@@ -57,7 +56,7 @@ parameters:
   buildDirectory: ''
 
 jobs:
-- job: ${{ coalesce(parameters.jobName, parameters.agentOs) }}
+- job: ${{ parameters.jobName }}
```
Comment (dougbu) [inline .azure/pipelines/jobs/default-build.yml:59]: Revert this change; it provides a fallback when the parameter isn't provided and doesn't hurt. May be needed in other pipelines for this branch.

File: .azure/pipelines/jobs/default-build.yml [hunk 5]
```diff
@@ -8,7 +8,7 @@
 #   poolName: string
 #       The name of the Azure DevOps agent pool to use.
 #   agentOs: string
-#       Used in templates to define variables which are OS specific. Typically from the set { Windows, Linux, macOS }
+#       Used in templates to define variables which are OS specific. Typically from the set { Windows, Linux, MacOS }
```
Comment (dougbu) [inline .azure/pipelines/jobs/default-build.yml:11]: Why change the case of `macOS`❔ Comparisons aren't case-sensitive.

**Review by dougbu — COMMENTED** [review]
Could you explain why the `run.*` scripts and build.ps1 needed to change❔

Separately, can we now delete [.azure/pipelines/pr-validation-temp.yml](https://github.com/dotnet/aspnetcore/blob/release/2.1/.azure/pipelines/pr-)validation-temp.yml)❔

File: .azure/pipelines/jobs/default-build.yml (hunk 5 above)
Comment (wtgodbe) [reply .azure/pipelines/jobs/default-build.yml:11]: Just for consistency, it was bothering me that some job names started with uppercase and some with lowercase so I did a find-replace.

**Comment by wtgodbe** [comment]
> Could you explain why the run.* scripts and build.ps1 needed to change?

I was duplicating what we have in extensions where we read `PB_` vars from the environment: https://github.com/dotnet/extensions/blob/3a9ef3fa2c14721f1803f7c89d82384438358330/build.ps1#L208-L216. However, this doesn't work for `PB_` vars which the pipeline defines as secrets, so I have another commit I'm testing now that goes back to having those vars be passed in from the command line. I'll have to do the same for extensions - it turns out that in that repo, we haven't been restoring from `PB_AdditionalRestoreSources` or importing `PB_PackageVersionProps` at all, since those variables are secrets and therefore don't actually get read from the command line.

File: .azure/pipelines/jobs/default-build.yml (hunk 4 above)
Comment (wtgodbe) [reply .azure/pipelines/jobs/default-build.yml:59]: Pushed a commit to my test branch that reverts this, will update this branch with the same change once I'm done testing

**Comment by wtgodbe** [comment]
This seems to be working now. Here's a build that was kicked by pipebuild: https://dnceng.visualstudio.com/internal/_build/results?buildId=933230&view=results

It pushed assets to this blob location, which looks the same as previous aspnetcore 2.1 blobs: 
https://dotnetfeed.blob.core.windows.net/orchestrated-release-2-1/20201223-10/aspnet/index.json

**Comment by dougbu** [comment]
> However, this doesn't work for `PB_` vars which the pipeline defines as secrets, so I have another commit I'm testing now that goes back to having those vars be passed in from the command line.

This isn't completely necessary. Adding `env:` settings in the YAML file, copying the secrets, would be more secure. (AzDO sometimes messes up masking secrets when logging command lines.)

**Comment by wtgodbe** [comment]
> This isn't completely necessary. Adding env: settings in the YAML file, copying the secrets, would be more secure. (AzDO sometimes messes up masking secrets when logging command lines.)

Oh, I didn't know that was more secure. I'll give that a try now.

**Comment by wtgodbe** [comment]
https://dnceng.visualstudio.com/internal/_build/results?buildId=933842&view=results

**Comment by wtgodbe** [comment]
Looks like that worked: https://dotnetfeed.blob.core.windows.net/orchestrated-release-2-1/20201229-10/aspnet/index.json

File: .azure/pipelines/jobs/default-build.yml (hunk 5 above)
Comment (JunTaoLuo) [reply .azure/pipelines/jobs/default-build.yml:11]: Interesting, I thought `macOS` is more canon than `MacOS` since that's how it's stylized by Apple

File: .azure/pipelines/jobs/default-build.yml [hunk 6]
```diff
@@ -85,20 +84,22 @@ jobs:
         queue: BuildPool.Server.Amd64.VS2017
   variables:
     AgentOsName: ${{ parameters.agentOs }}
-    ASPNETCORE_TEST_LOG_MAXPATH: "200" # Keep test log file name length low enough for artifact zipping
-    DOTNET_HOME: $(Agent.BuildDirectory)/.dotnet
     BuildScript: ${{ parameters.buildScript }}
-    BuildScriptArgs: ${{ parameters.buildArgs }}
     BuildConfiguration: ${{ parameters.configuration }}
     BuildDirectory: ${{ parameters.buildDirectory }}
-    TeamName: AspNetCore
```
Comment (JunTaoLuo) [inline .azure/pipelines/jobs/default-build.yml:94]: Did we want to start capturing binlogs all the time? Wouldn't this use up a lot of storage?

**Review by JunTaoLuo — COMMENTED** [review]
First batch of comments

File: .azure/pipelines/jobs/default-build.yml (hunk 6 above)
Comment (wtgodbe) [reply .azure/pipelines/jobs/default-build.yml:94]: I didn't think it was too impactful for 2.1 since we rarely build that branch - but, I could turn it off for public builds like we do in other branches

File: .azure/pipelines/jobs/default-build.yml (hunk 5 above)
Comment (wtgodbe) [reply .azure/pipelines/jobs/default-build.yml:11]: Didn't know that - I can always change it back, not too big of a deal

File: src/Installers/Windows/AspNetCoreModule-Setup/build/settings/common.props (hunk 2 above)
Comment (JunTaoLuo) [reply src/Installers/Windows/AspNetCoreModule-Setup/build/settings/common.props:19]: Maybe @jkotalik knows something about this? Looks like the change was added in https://github.com/dotnet/aspnetcore/commit/8356baf7a612bd21cc92f8fc7098e292274de102 but that commit is huge.

File: src/Installers/Windows/Wix.props (hunk 3 above)
Comment (JunTaoLuo) [reply src/Installers/Windows/Wix.props:4]: I think this is probably the best we can do for now but I'm a little concerned about the modulo logic since it can eventually silently cause issues. Instead of modulo maybe just use subtraction and error out if our build numbers are too large so we would be notified and figure something else out at that time? Though I don't know if that will ever happen, I feel it's safer since it'll fail loudly.

File: build/repo.targets [hunk 7]
```diff
@@ -274,4 +274,63 @@
...
+      <!-- Make sure we're not disabling code signing -->
+      <DisableCodeSigning>false</DisableCodeSigning>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <!-- We're passing a custom list of files - clear out the exclusion list so it doesn't interfere -->
+      <FilesToExcludeFromSigning Remove="@(FilesToExcludeFromSigning)" />
+
+      <!-- Make sure 3rd party binaries get 3rd party certificate -->
+      <CustomFileSignInfo Include="MsgPack.dll" CertificateName="3PartySHA2" />
```
Comment (JunTaoLuo) [inline build/repo.targets:290]: How was this list generated? I just want to double check but it looks fine.

**Review by JunTaoLuo — COMMENTED** [review]
2nd batch

File: .azure/pipelines/ci.yml [hunk 8]
```diff
@@ -7,22 +7,716 @@ pr:
...
+      displayName: Build win-x64 SharedFX
+    - powershell: eng\scripts\KillProcesses.ps1
+      displayName: Kill processes
+      condition: always()
+    - script: .\build.cmd
+              -ci
+              /p:SignType=$(_SignType)
+              $(BuildScriptArgs)
+              $(SharedFxArgs)
+              /p:SharedFxRID=win-x86
```
Comment (JunTaoLuo) [inline .azure/pipelines/ci.yml:118]: Hmm the display name says x64 but this RID says x86? I think the name got flipped with the one below this.

File: build/repo.targets (hunk 7 above)
Comment (wtgodbe) [reply build/repo.targets:290]: I manually looked at the contents of all of the .nupkg's/.zips that we're signing, and cross-referenced with what's in master: https://github.com/dotnet/aspnetcore/blob/master/eng/Signing.props

File: src/Installers/Windows/Wix.props (hunk 3 above)
Comment (wtgodbe) [reply src/Installers/Windows/Wix.props:4]: I think the only risk is if we produce two versions that are exactly the same, which would happen if 2 builds of the same patch were exactly 65,535 AzDO builds apart (which seems unlikely). I believe 2.1.25.1 would still be considered "higher" than 2.1.24.65535 by WiX - @joeloff does that sound right? That is, it isn't strictly an issue if the `buildNumbers` aren't always increasing?

File: .azure/pipelines/ci.yml [hunk 9]
```diff
@@ -7,22 +7,716 @@ pr:
...
+      continueOnError: true
+      inputs:
+        pathtoPublish: .deps/dependencies.g.props
+        artifactName: artifacts-dependencies-props
+        artifactType: Container
+        parallel: true
+
+  - job: MacOs_SharedFx
+    displayName: Build Osx-x64 SharedFx
+    dependsOn: Windows_Build
```
Comment (JunTaoLuo) [inline .azure/pipelines/ci.yml:170]: Why is this depending on windows build?

File: src/Installers/Windows/Wix.props (hunk 3 above)
Comment (wtgodbe) [reply src/Installers/Windows/Wix.props:4]: This _would_ be an issue if we were producing previews of this branch, because there wouldn't be a guarantee that the versions would always be increasing (e.g. preview1's `BuildNumber` could be higher than preview2's, and they'd have the same patch number)

**Review by JunTaoLuo — COMMENTED** [review]
Last batch, so far so good.

File: .azure/pipelines/ci.yml (hunk 9 above)
Comment (wtgodbe) [reply .azure/pipelines/ci.yml:170]: The Unix SharedFx builds depend on the windows product build in TeamCity as well - they download windows artifacts to `.deps` and reference them from there

File: .azure/pipelines/ci.yml (hunk 8 above)
Comment (wtgodbe) [reply .azure/pipelines/ci.yml:118]: The display name for this step is below, on line 125. The display name on line 109 is for the first build.cmd invocation

**Review by JunTaoLuo — APPROVED** [review]
